File: apps/api/app/services/ai/providers/openai.rb
# frozen_string_literal: true

module Ai
  module Providers
    # OpenAI provider using the official Ruby SDK's Responses API with
    # structured outputs. The response is parsed into an Ai::Schemas::TicketAnalysis
    # BaseModel instance, guaranteeing the JSON conforms to the defined schema.
    #
    # Configuration:
    #   ENV["OPENAI_API_KEY"]           - required, raises KeyError if missing
    #   ENV["OPENAI_MODEL"]             - optional, defaults to "gpt-5.6-luna"
    #   ENV["OPENAI_CONNECT_TIMEOUT"]   - connection timeout in seconds (default: 5)
    #   ENV["OPENAI_READ_TIMEOUT"]      - read/request timeout in seconds (default: 30)
    #
    # A new client is instantiated per call to avoid stale connections.
    # Refusal responses from the model are filtered out; if no valid
    # content remains, a RuntimeError is raised.
    class Openai
      SYSTEM_PROMPT = <<~PROMPT
        You are a support ticket analyst. Analyze the given customer support ticket and extract structured insights.

        For each ticket, determine:
        - sentiment: The overall customer sentiment (positive, negative, neutral, or mixed)
        - summary: A concise 1-2 sentence summary of the ticket's core issue or request
        - category: Exactly one of: billing, technical_issue, feature_request, account, onboarding, integrations, performance, general
        - confidence: Your confidence in this analysis from 0.0 to 1.0
        - feature_request: Whether this ticket contains or implies a feature request
        - bug_report: Whether this ticket reports a bug or defect
        - knowledge_gap: Whether this ticket reveals a gap in documentation or user knowledge

        Be precise and objective. Base your analysis only on the ticket content provided.
      PROMPT

      class << self
        def analyze(ticket:)
          response = client.responses.create(
            model: model,
            input: build_input(ticket),
            text: Ai::Schemas::TicketAnalysis
          )

          extract_result(response)
        end

        def generate_executive_summary(input:)
          response = client.responses.create(
            model: model,
            input: input,
            text: Ai::Schemas::ExecutiveSummaryOutput
          )

          extract_result(response)
        end

        def detect_themes(input:)
          response = client.responses.create(
            model: model,
            input: input,
            text: Ai::Schemas::ThemeDetectionOutput
          )

          extract_result(response)
        end

        # Generate embeddings for an array of text strings.
        #
        # Returns an array of embedding vectors in the same input order.
        # Uses the OpenAI embeddings API with explicit dimensions.
        #
        # Options:
        #   model:         embedding model (default: "text-embedding-3-small")
        #   dimensions:    vector dimensions (default: 1536)
        #   include_usage: when true, returns { vectors:, usage: { prompt_tokens: } }
        def embed(texts:, model: "text-embedding-3-small", dimensions: 1536, include_usage: false)
          response = with_retry do
            client.embeddings.create(
              model: model,
              input: texts,
              dimensions: dimensions
            )
          end

          # Sort by index to guarantee input order
          vectors = response.data.sort_by(&:index).map(&:embedding)

          if include_usage
            { vectors: vectors, usage: { prompt_tokens: response.usage&.prompt_tokens } }
          else
            vectors
          end
        end

        # Send a chat completion request.
        #
        # messages: array of { role:, content: } hashes
        # model:    model name (defaults to ENV["OPENAI_MODEL"] / "gpt-5.6-luna")
        #
        # Returns the full ChatCompletion response object.
        def chat(messages:, model: self.model)
          with_retry do
            client.chat.completions.create(
              model: model,
              messages: messages
            )
          end
        end

        # Classify an OpenAI error as retryable or permanent.
        #
        # Retryable: timeouts, rate limits, server errors (5xx).
        # Permanent: authentication, malformed requests, client errors.
        def retryable_error?(error)
          case error
          when OpenAI::Errors::APITimeoutError, OpenAI::Errors::APIConnectionError
            true
          when OpenAI::Errors::RateLimitError, OpenAI::Errors::InternalServerError
            true
          else
            false
          end
        end

        private

        MAX_RETRIES = ENV.fetch("OPENAI_MAX_RETRIES", "2").to_i
        BASE_DELAY = 0.5 # seconds

        # Bounded retry with exponential backoff for transient failures.
        # Only retries errors classified as retryable.
        # Permanent errors are re-raised immediately.
        def with_retry
          attempts = 0
          begin
            attempts += 1
            yield
          rescue OpenAI::Errors::APIError => e
            raise unless retryable_error?(e)
            raise if attempts > MAX_RETRIES

            delay = BASE_DELAY * (2**(attempts - 1))
            Rails.logger.warn(
              "[OpenAI] Retryable error (attempt #{attempts}/#{MAX_RETRIES + 1}): " \
              "#{e.class.name} — retrying in #{delay}s"
            )
            sleep(delay)
            retry
          end
        end

        def client
          OpenAI::Client.new(
            api_key: api_key,
            timeout: {
              connect: connect_timeout,
              read: read_timeout
            }
          )
        end

        def api_key
          ENV.fetch("OPENAI_API_KEY") { raise KeyError, "OPENAI_API_KEY environment variable is not set" }
        end

        def model
          ENV.fetch("OPENAI_MODEL", "gpt-5.6-luna")
        end

        def connect_timeout
          ENV.fetch("OPENAI_CONNECT_TIMEOUT", "5").to_i
        end

        def read_timeout
          ENV.fetch("OPENAI_READ_TIMEOUT", "30").to_i
        end

        def build_input(ticket)
          user_content = <<~CONTENT
            Subject: #{ticket.subject}
            Description: #{ticket.description}
          CONTENT

          [
            { role: :system, content: SYSTEM_PROMPT },
            { role: :user, content: user_content }
          ]
        end

        def extract_result(response)
          content = response
            .output
            .flat_map(&:content)
            .grep_v(OpenAI::Models::Responses::ResponseOutputRefusal)
            .first

          raise "No valid response content returned from OpenAI" unless content&.parsed

          content.parsed
        end
      end
    end
  end
end

File: apps/ai/app/providers/openai.py
import logging
import time

from openai import OpenAI
from openai import AuthenticationError as OpenAIAuthError
from openai import RateLimitError as OpenAIRateLimitError
from openai import APITimeoutError as OpenAITimeoutError
from openai import APIConnectionError as OpenAIConnectionError
from openai import APIStatusError as OpenAIStatusError

from app.config import settings
from app.schemas import TicketAnalysisResponse

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """\
You are a support ticket analyst. Analyze the given customer support ticket and extract structured insights.

For each ticket, determine:
- sentiment: The overall customer sentiment (positive, negative, neutral, or mixed)
- summary: A concise 1-2 sentence summary of the ticket's core issue or request
- category: Exactly one of: billing, technical_issue, feature_request, account, onboarding, integrations, performance, general
- confidence: Your confidence in this analysis from 0.0 to 1.0
- feature_request: Whether this ticket contains or implies a feature request
- bug_report: Whether this ticket reports a bug or defect
- knowledge_gap: Whether this ticket reveals a gap in documentation or user knowledge

Be precise and objective. Base your analysis only on the ticket content provided."""


class AIServiceError(Exception):
    """Base error for AI service failures."""


class ConfigurationError(AIServiceError):
    """Missing or invalid configuration."""


class ProviderError(AIServiceError):
    """Base for provider-specific errors."""


class AuthenticationError(ProviderError):
    """OpenAI authentication failure."""


class RateLimitError(ProviderError):
    """OpenAI rate limit exceeded."""


class TimeoutError(ProviderError):
    """OpenAI request timeout."""


class APIError(ProviderError):
    """OpenAI API/server failure."""


class InvalidResponseError(AIServiceError):
    """Model returned an unparseable or invalid response."""


class RefusalError(AIServiceError):
    """Model refused to analyze the content."""


def analyze_ticket(ticket_id: str, subject: str, description: str) -> TicketAnalysisResponse:
    """Call OpenAI to analyze a support ticket and return structured analysis."""
    if not settings.openai_api_key:
        raise ConfigurationError("OPENAI_API_KEY is not configured")

    client = OpenAI(api_key=settings.openai_api_key)

    user_content = f"Subject: {subject}\nDescription: {description}"

    start_time = time.monotonic()
    try:
        response = client.responses.parse(
            model=settings.openai_model,
            input=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": user_content},
            ],
            text_format=TicketAnalysisResponse,
        )
    except OpenAIAuthError as e:
        logger.error("ticket_id=%s provider=openai error=auth_failure", ticket_id)
        raise AuthenticationError("OpenAI authentication failed") from e
    except OpenAIRateLimitError as e:
        logger.warning("ticket_id=%s provider=openai error=rate_limited", ticket_id)
        raise RateLimitError("OpenAI rate limit exceeded") from e
    except OpenAITimeoutError as e:
        logger.warning("ticket_id=%s provider=openai error=timeout", ticket_id)
        raise TimeoutError("OpenAI request timed out") from e
    except (OpenAIConnectionError, OpenAIStatusError) as e:
        logger.error("ticket_id=%s provider=openai error=api_error", ticket_id)
        raise APIError("OpenAI API error") from e

    duration_ms = int((time.monotonic() - start_time) * 1000)

    # Extract parsed content, handling refusals
    output_content = None
    for item in response.output:
        for content_block in item.content:
            if hasattr(content_block, "refusal") and content_block.refusal:
                logger.warning("ticket_id=%s provider=openai error=refusal", ticket_id)
                raise RefusalError("Model refused to analyze the ticket")
            if hasattr(content_block, "parsed") and content_block.parsed is not None:
                output_content = content_block.parsed
                break
        if output_content is not None:
            break

    if output_content is None:
        logger.error("ticket_id=%s provider=openai error=invalid_response", ticket_id)
        raise InvalidResponseError("No valid structured response from model")

    logger.info(
        "ticket_id=%s provider=openai model=%s duration_ms=%d status=success",
        ticket_id,
        settings.openai_model,
        duration_ms,
    )

    # The SDK returns a TicketAnalysisResponse instance via structured outputs.
    # Override ticket_id to match the request (model doesn't know it).
    return TicketAnalysisResponse(
        ticket_id=ticket_id,
        sentiment=output_content.sentiment,
        summary=output_content.summary,
        category=output_content.category,
        confidence=output_content.confidence,
        feature_request=output_content.feature_request,
        bug_report=output_content.bug_report,
        knowledge_gap=output_content.knowledge_gap,
    )
